app/utils/YarnLockRepositoryParser.scala

```diff
@@ -12,7 +12,8 @@ import scala.util.matching.Regex
 object YarnLockRepositoryParser extends AbstractParser {
 
   private val buildFileName = "yarn.lock"
-  private val yarnVersion = """"?([^\n,]*)@.*:\n\s+version "(.*)"""".r
+  private val yarnVersion = """"?([^\n,]*)@(.*):\n\s+version "(.*)"""".r
```

Comment:
That's not going to work in all cases, unfortunately. Also, the regex might prone to backtracking.

I have a project with
- `package.json`
```
"dependencies": {
  // …
  "moment": "2.24.0",
  //…
}
```
- `yarn.lock`
```
moment@2.24.0, moment@^2.24.0:
  version "2.24.0"
  resolved "https://nexus.wired-aws.com/repository/npm-group/moment/-/moment-2.24.0.tgz#0d055d53f5052aa653c9f6eb68bb5d12bf5c2b5b"
  integrity sha512-bV7f+6l2QigeBBZSM/6yTNq4P2fNpSWj/0e7jQcy87A8e7o2nAfP/34/2ky5Vw4B9S446EtIhodAzkFCcR4dQg==

"moment@>= 2.9.0":
  version "2.22.2"
  resolved "https://registry.yarnpkg.com/moment/-/moment-2.22.2.tgz#3c257f9839fc0e93ff53149632239eb90783ff66"
```

The correct dependency is of course the first, but with the regex, it will be split as `moment` + `2.24.0, moment@^2.24.0`, which is obviously not the declared version.

Reply:
Ah, actually it will work because you've used `contains()`, but isn't it a bit too broad?

Reply:
Yes, it's true. Some semver are tricky to match like the or operator (ex `"y18n@^3.2.1 || ^4.0.0", y18n@^4.0.0:`) so it's difficult to be more specific.
I cannot find a case where this solution doesnt work (the same version would have to be in two ranges without intersection)

Reply:
some cases like `1.0` and `1.0.1` can be problematic

Reply:
@fpavageau i made some changes to fix the case where a version is contained in an other.